cmd/sops/edit.go

```diff
@@ -0,0 +1,274 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+
+	"crypto/md5"
+	"io"
+	"os/exec"
+	"strings"
+
+	"path"
+
+	"bufio"
+	"bytes"
+
+	"github.com/google/shlex"
+	"go.mozilla.org/sops"
+	"go.mozilla.org/sops/cmd/sops/codes"
+	"go.mozilla.org/sops/cmd/sops/common"
+	"go.mozilla.org/sops/keyservice"
+	"go.mozilla.org/sops/stores/json"
+	"gopkg.in/urfave/cli.v1"
+)
+
+type editOpts struct {
+	Cipher         sops.DataKeyCipher
+	InputStore     sops.Store
+	OutputStore    sops.Store
+	InputPath      string
+	IgnoreMAC      bool
+	KeyServices    []keyservice.KeyServiceClient
+	ShowMasterKeys bool
+}
+
+type editExampleOpts struct {
+	editOpts
+	UnencryptedSuffix string
+	KeyGroups         []sops.KeyGroup
+	GroupQuorum       int
+}
+
+var exampleTree = sops.TreeBranch{
+	sops.TreeItem{
+		Key:   "hello",
+		Value: `Welcome to SOPS! Edit this file as you please!`,
+	},
+	sops.TreeItem{
+		Key:   "example_key",
+		Value: "example_value",
+	},
+	sops.TreeItem{
+		Key: "example_array",
+		Value: []interface{}{
+			"example_value1",
+			"example_value2",
+		},
+	},
+	sops.TreeItem{
+		Key:   "example_number",
+		Value: 1234.56789,
+	},
+	sops.TreeItem{
+		Key:   "example_booleans",
+		Value: []interface{}{true, false},
+	},
+}
+
+type runEditorUntilOkOpts struct {
+	TmpFile        *os.File
+	OriginalHash   []byte
+	InputStore     sops.Store
+	ShowMasterKeys bool
+	Tree           *sops.Tree
+}
+
+func EditExample(opts editExampleOpts) ([]byte, error) {
+	// Load the example file
+	var fileBytes []byte
+	if _, ok := opts.InputStore.(*json.BinaryStore); ok {
+		// Get the value under the first key
+		fileBytes = []byte(exampleTree[0].Value.(string))
+	} else {
+		var err error
+		fileBytes, err = opts.InputStore.Marshal(exampleTree)
+		if err != nil {
+			return nil, err
+		}
+	}
+	var tree sops.Tree
+	branch, err := opts.InputStore.Unmarshal(fileBytes)
+	if err != nil {
+		return nil, cli.NewExitError(fmt.Sprintf("Error unmarshalling file: %s", err), codes.CouldNotReadInputFile)
+	}
+	tree.Branch = branch
+	tree.Metadata = sops.Metadata{
+		KeyGroups:         opts.KeyGroups,
+		UnencryptedSuffix: opts.UnencryptedSuffix,
+		Version:           version,
+		ShamirQuorum:      opts.GroupQuorum,
+	}
+
+	// Generate a data key
+	dataKey, errs := tree.GenerateDataKeyWithKeyServices(opts.KeyServices)
+	if len(errs) > 0 {
+		return nil, cli.NewExitError(fmt.Sprintf("Error encrypting the data key with one or more master keys: %s", errs), codes.CouldNotRetrieveKey)
+	}
+	stash := make(map[string][]interface{})
+
+	return edit(opts.editOpts, &tree, dataKey, stash)
+}
+
+func Edit(opts editOpts) ([]byte, error) {
```

Comment:
These are newly exported functions, and I'm not clear on the value of exporting them. Care to clarify?